### PR TITLE
Add actions functionality

### PR DESCRIPTION
**Pull request by proof666**
**Please confirm you have linted your code and fixed any issues:**

* [x] I ran `yarn run fix` on my PR and fixed any formatting issues

**Please provide information on what your PR achieves and any issues that it addresses:**

Hi, @jemise111! 

I know you have example of swipe to delete functionality but this approach seems to me not obvious and limited. 
This changes are introduce new functionality to make it easier to work with features such as swipe to delete or swipe to cart (see #254 for reference). Also it makes easier to change the UI based on swipe state.
I find it easier to use with less effort and code. No need to manage a list of animated values for all rows. Behavior can be controlled more granular.  
Also one of my requirements is the reaction of the hidden component when the action is activated (i.e. expand delete button to full width). Without this changes I did not find a way to achieve this without rendering the whole list after changing state (which potentially led to performance problems on large lists).

So this is my solution... 

Added configuration props:
`leftActivationValue` - TranslateX value for firing `onLeftActionStatusChange` (positive number)
`rightActivationValue` - TranslateX value for firing `onRightActionStatusChange` (negative number)
`leftActionValue` - TranslateX value for left action to which the row will be shifted after gesture release
`rightActionValue` - TranslateX value for right action to which the row will be shifted after gesture release

Added handlers:
`onLeftAction` - fired when row shifted to `leftActivationValue`
`onRightAction` - fired when row shifted to `rightActivationValue`
`onLeftActionStatusChange` - fired once when swipe value reached the `leftActivationValue`
`onRightActionStatusChange` - fired once when swipe value reached the `rightActivationValue`

Added props to Visible and Hidden components (boolean values):
`swipeAnimatedValue` - swipeRow `_translateX`
`leftActionActivated`
`rightActionActivated`
`leftActionEvaluated`
`rightActionEvaluated`
...currently `*ActionEvaluated` is changed to true only once and never resets. Need to discuss whether to reset the value and in which moment.

I modified example application with new functionality to make it clearer why these changes are needed and how to use them.
screen

If you like these changes and they do not contradict your vision, I will add documentation and type definitions. Maybe some additional handlers such as *didActivated or *willActivated... or useSwipeRowState hook.

Please give your feedback.

### CODE REVIEW

File: components/SwipeRow.js [hunk 1]
```diff
@@ -347,27 +423,68 @@ class SwipeRow extends Component {
...
+            action = () => {
+                this.props.onLeftAction && this.props.onLeftAction();
+                !this.state.leftActionEvaluated &&
+                    this.setState({
+                        leftActionEvaluated: true,
+                    });
+            };
+        }
+
+        this.manuallySwipeRow(toValue, action);
```
Comment (proof666) [inline components/SwipeRow.js]: @jemise111, the default animation type is `spring` now... What do you think about the ability to set the type of animation and its settings as props (i.e. animationType, animationConfig)?

**Comment by jemise111** [comment]
Hey @proof666 thanks for putting in the time to make this! Very excited to look through it. Things are a little busy for me at work right now so don't think I forgot about this if I don't get to it in the next couple days. I will look through it asap and give you feedback

File: components/SwipeRow.js (hunk 1 above)
Comment (jemise111) [reply components/SwipeRow.js]: Absolutely, as long as you keep the defaults as they are now I'm on board

File: components/SwipeRow.js [hunk 2]
```diff
@@ -326,6 +384,17 @@ class SwipeRow extends Component {
                         ? 0
                         : this.props.leftOpenValue;
                 }
+                if (
+                    this.currentTranslateX - projectedExtraPixels >
+                    this.props.leftActivationValue
+                ) {
+                    // we're more than halfway
```
Comment (jemise111) [inline components/SwipeRow.js]: I think this comment is misleading - I think it should say something like "We've passed the threshold to trigger the leftActionValue" .. something like that

File: components/SwipeRow.js [hunk 3]
```diff
@@ -354,20 +432,61 @@ class SwipeRow extends Component {
                         ? 0
                         : this.props.rightOpenValue;
                 }
+                if (
+                    this.currentTranslateX - projectedExtraPixels <
+                    this.props.rightActivationValue
+                ) {
+                    // we're more than halfway
```
Comment (jemise111) [inline components/SwipeRow.js]: Same comment here

**Comment by jemise111** [comment]
Hey @proof666 really amazing work here. The PR is very thorough and I played around with the example and everything seems to be working really nicely. I made a few small comments to address.

To respond to your statement about "they do not contradict your vision.." I mean personally I like managing animated values in an Object at the highest level. I feel like I ultimately have control over everything at once. And it I wanted to do some form of stagger animation using multiple values i would have the ability to do so. That being said this approach seems a bit more user friendly and you've clearly given this a lot of thought so I would love to get these changes in the library as well and give the end user the option to manage animated values and swipe actions in either way.

In addition to the changes requested my other big ask would be that the `handlePanResponderEnd` function in `SwipeRow.js` is getting a little bloated (this is very much my fault). Could we move at least the logic at the end of the function (lines 469 - 488) out into a separate function. Totally optional but if it is possible that would be awesome.

My last request would be (in addition to the code you added to `example.js`) to create a new doc in the `docs` folder and do a small writeup of how the feature works and how to use it. 

After that's all done I will get this merged in released. Maybe call it `3.0.0` since it's a nice new set of features.

Please ask any questions and thanks so much!

**Comment by proof666** [comment]
Thank you for your time, it is very cool that the proposed changes can be accepted. I agree with all the comments and will try to bring the pull request to the final state in the near future. In addition, I already managed to find a couple of non-critical bugs. Now I have at work the time for the next release and time is running out. But next week I’ll try to finish it all.

**Comment by jemise111** [comment]
@proof666 Sounds good!

File: components/SwipeRow.js (hunk 1 above)
Comment (proof666) [reply components/SwipeRow.js]: I will propose this changes in separate PR

**Comment by proof666** [comment]
@jemise111, sorry, I'm a little busy at work right now. It remains to prepare the documentation for examples. I will try to complete the work soon.

**Comment by jemise111** [comment]
@proof666 no worries, I know the feeling :)

**Comment by proof666** [comment]
Hi @jemise111!
Sorry to take so long, but I was very busy at work.
During this time, I made changes in accordance with your previous review:
- [x] small refactoring of bloated `handlePanResponderEnd` function
- [x] create a new doc in the docs with Actions descriptions
- [x] all inline comments resolved

And:
- [x] fix some bugs in Actions logic and add missing types
- [x] rename `*actionEvaluated`  props to `*actionState`. Make `*actionState` toggle-like (switch true\false on each change) and allow to set initial state for the row.
- [x] update example (in order to respond to files structure changes)
- [x] update other docs (where it is applicable)

Pay attention:
- The formatting of the tables in the markdown documents has changed. So diffs looks ugly. I did not immediately pay attention to this. If it is critical, please inform me, i will try to fix it.
- I am not a native speaker, so feel free to make any changes to the description and documentation.
- You probably need to refresh the gif on the main page again. But now there are 7 sections and it doesn’t look so elegant as now.

PS: FYI, I no longer plan to make any changes to this pull request except changes based on the results of the review

**Comment by jemise111** [comment]
Hey @proof666 so sorry I caused you merge conflicts! I wasn't thinking and merged a bunch of PRs before having a look at this one. I will review this tonight or tomorrow and will try to resolve this conflicts on my own if I can manage, thanks for your patience!

**Comment by jemise111** [comment]
@proof666 I did some testing and found a piece of functionality that broke. If you run the example "SwipeValueBasedUI" you'll notice something is wrong with the `onSwipeValueChange` prop. Would you mind having a look? Thanks!

File: SwipeListExample/examples/actions.js [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+import SwipeListView from '../SwipeListView';
+
+export default function Actions() {
+    const [listData, setListData] = useState(
+        Array(20)
+            .fill('')
+            .map((_, i) => ({
+                key: `${i}`,
+                text: `item #${i}`,
+                initialLeftActionState: i % 2,
```
Comment (jemise111) [inline SwipeListExample/examples/actions.js]: I think this one should be a bool? Maybe you mean to write `i % 2 !== 0`?

**Comment by proof666** [comment]
Hey @jemise111, 

A brief answers:

1. The link to Actions doc already added to main readme (https://github.com/proof666/react-native-swipe-list-view/blob/master/README.md). Did you mean something else?
Screenshot 2020-03-30 at 11 25 35

I will change "Swipe to Delete" to "Swipe to Delete (also see Actions for an alternative way to achieve this)" as you suggest.

2. Yes, it is a formatter. I will fix it.

3. Ok :)

I'll see what's wrong with SwipeValueBasedUI example today

I completely agree that it’s better to release these changes under `rc` for some time.

File: SwipeListExample/examples/actions.js (hunk 4 above)
Comment (proof666) [reply SwipeListExample/examples/actions.js]: Yep, i think so. I will fix it.

Comment (proof666) [reply SwipeListExample/examples/actions.js]: Fixed

**Comment by proof666** [comment]
> @proof666 I did some testing and found a piece of functionality that broke. If you run the example "SwipeValueBasedUI" you'll notice something is wrong with the `onSwipeValueChange` prop. Would you mind having a look? Thanks!

I cant see any problems with `onSwipeValueChange` in "SwipeValueBasedUI" on https://github.com/proof666/react-native-swipe-list-view. It just work as expected, icon scales when i swipe the row. No errors in console.
image

Could you provide more detailed repro steps or error?

**Comment by proof666** [comment]
@jemise111 
I am very glad that the work on these changes is completed and they are accepted. Thank you for your responsiveness and interest. 

Feel free to mention me in related issues.

**Pull request merged**